rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,264 @@
+Proposal for updating quantization scale functionality in oneDNN
+================================================================
+
+## 0. Motivation
+
+There are multiple requests for extending scaling support in oneDNN primitives:
+- run-time scaling in Eltwise post-op;
+- scaling for src1 in Binary post-op;
+- scaling for inputs in Concat.
+
+All of these requests can be addressed separately, but taking into account
+already existing oneDNN mechanisms can help to avoid duplications and to unify
+scaling in the future.
+
+
+## 1. Introduction
+
+This chapter provides a comprehensive overview of scaling support in oneDNN.
+
+
+### 1.1. Quantization
+
+8-bit Inference development covers more and more deep learning models, which
+means more oneDNN primitives are required to support 8-bit data type. As
+explained before [[#1]][1] quantization requires scale and shift. This RFC only
+covers scale.
+
+
+### 1.2. Scaling
+
+Scaling is required in the following situations:
+ - Output scaling. An intermediate result of an operation should be re-quantized
+ to fit into a destination data type. For example:
+ ```
+ dst_f32 = conv(src_f32, wei_f32)
+
+ src_f32 = src_s8 * src_scale_f32
+ wei_f32 = wei_s8 * wei_scale_f32
+ dst_f32 = dst_s8 * dst_scale_f32
+
+ output_scale_f32 = src_scale_f32 * wei_scale_f32 / dst_scale_f32
+ acc_s32 = conv(src_s8, wei_s8)
+ dst_s8 = acc_s32 * output_scale_f32
+ ```
+ - Argument scaling. Operation's inputs have different scales and they should
+ be taken into account while computing the operation. For example:
+ ```
+ dst_f32 = src0_f32 + src1_f32
+
+ src0_f32 = src0_s8 * src0_scale_f32
+ src1_f32 = src1_s8 * src1_scale_f32
+ dst_f32 = dst_s8 * dst_scale_f32
+
+ # scales src0_scale_f32 and src1_scale_f32 are required to compute operation
+ correctly.
+ acc_s32 = src0_s8 + src1_s8 # doesn't work if scales are different
+ dst_s8 = acc_s32 * scale_f32
+ ```
+
+The following features make scaling more flexible:
+- runtime support. Allows model to set scale at runtime;
+- broadcasting/mask. Allows model to use multiple scales.
+
+
+### 1.3. Quantization scaling in oneDNN
+
+oneDNN has multiple mechanisms to support output and argument scaling:
+
+| mechanism                      | argument  | compile-time | run-time | mask |
+|--------------------------------|-----------|--------------|----------|------|
+| attr::output_scale             | dst       | yes          | yes      | yes  |
+| attr::scales                   | all       | yes          | yes      | yes  |
+| attr::p_op::binary_mul         | dst       | no           | yes      | yes  |
+| attr::rnn_data_qparams         | src & dst | yes          | no       | no   |
+| attr::rnn_weights_qparams      | wei       | yes          | no       | no   |
+| attr::p_op::built-in parameter | dst       | yes          | no       | no   |
+
+
+The following table shows scaling mechanisms support by primitives:
```

Comment:
I feel this the support status of primitive APIs, not primitive implementations. We know that CPU and GPU have different support status.

Reply:
Yes, correct, fixed it. Thanks for catching this.